src/thread.rs

```diff
@@ -0,0 +1,583 @@
+//! Implements threads.
+
+use std::cell::RefCell;
+use std::convert::TryFrom;
+use std::num::NonZeroU32;
+
+use log::trace;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def_id::DefId;
+use rustc_index::vec::{Idx, IndexVec};
+use rustc_middle::{
+    middle::codegen_fn_attrs::CodegenFnAttrFlags,
+    mir,
+    ty::{self, Instance},
+};
+
+use crate::*;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum SchedulingAction {
+    /// Execute step on the active thread.
+    ExecuteStep,
+    /// Execute destructors of the active thread.
+    ExecuteDtors,
+    /// Stop the program.
+    Stop,
+}
+
+/// A thread identifier.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct ThreadId(usize);
+
+/// The main thread. When it terminates, the whole application terminates.
+const MAIN_THREAD: ThreadId = ThreadId(0);
+
+impl Idx for ThreadId {
+    fn new(idx: usize) -> Self {
+        ThreadId(idx)
+    }
+    fn index(self) -> usize {
+        self.0
+    }
+}
+
+impl From<u64> for ThreadId {
+    fn from(id: u64) -> Self {
+        Self(usize::try_from(id).unwrap())
+    }
+}
+
+impl From<u32> for ThreadId {
+    fn from(id: u32) -> Self {
+        Self(usize::try_from(id).unwrap())
+    }
+}
+
+impl ThreadId {
+    pub fn to_u32_scalar<'tcx>(&self) -> Scalar<Tag> {
+        Scalar::from_u32(u32::try_from(self.0).unwrap())
+    }
+}
+
+/// An identifier of a set of blocked threads.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct BlockSetId(NonZeroU32);
+
+impl From<u32> for BlockSetId {
+    fn from(id: u32) -> Self {
+        Self(NonZeroU32::new(id).expect("0 is not a valid blockset id"))
```

Comment:
`From` conversions should usually be infallibe; `TryFrom` is for fallible conversions.

Here I feel like mirroring `NonZeroU32` API would be better: have `BlockSetId::new` that returns `Option<BlockSetid>`. Or maybe let it do the `unwrap` already; fallible `new` seems fine.